lighthouse-core/report/html/components/gauge.js

```diff
@@ -0,0 +1,398 @@
+        <circle class="lh-gauge-base" r="56" cx="60" cy="60"></circle>
+        <circle class="lh-gauge-arc" transform="rotate(-90 60 60)" r="56" cx="60" cy="60"></circle>
+      </svg>
+    </div>
+    <div class="lh-gauge__percentage"></div>
+    <div class="lh-gauge__label"></div>
+  </a>
+`;
+
+const attributes = {
+  score: '0.5',
+  variant: 'pass',
+  label: 'Lighthouse',
+};
+
+class LighthouseGaugeElement extends HTMLElement {
+  constructor() {
+    super();
+    this._shadowRoot = this.attachShadow({mode: 'open'});
+    this._shadowRoot.appendChild(template.content.cloneNode(true));
+    this._attributes = {...attributes};
+
+    for (const key in Object.keys(attributes)) {
+      const attributeName = /** @type {keyof typeof attributes} */ (key);
+      Object.defineProperty(this, attributeName, {
+        get: () => this._attributes[attributeName],
+        set: value => this.setAttribute(attributeName, String(value)),
+      });
+    }
+
+    this._render();
+  }
+
+  static get observedAttributes() {
+    return Object.keys(attributes);
+  }
+
+
+  connectedCallback() {
+    for (const attribute of this.attributes) {
+      // @ts-ignore
+      this[attribute.name] = attribute.nodeValue;
+    }
+  }
+
+  /** @param {string} name @param {string} oldValue @param {string} newValue */
+  attributeChangedCallback(name, oldValue, newValue) {
+    if (!(name in attributes) || oldValue === newValue) return;
+
+    // @ts-ignore
+    this._attributes[name] = newValue;
+    this._render();
+  }
+
+  _render() {
+    const wrapper = this._shadowRoot.querySelector('.lh-gauge__wrapper');
+    if (!wrapper) throw new Error('Impossible');
+    wrapper.classList.add(`lh-gauge__wrapper--${this._attributes.variant}`);
+
+    // Cast `null` to 0
+    const numericScore = Number(this._attributes.score) || 0;
+    const gauge = this._shadowRoot.querySelector('.lh-gauge');
+    if (!gauge) throw new Error('Impossible');
+
+    // 352 is ~= 2 * Math.PI * gauge radius (56)
+    // https://codepen.io/xgad/post/svg-radial-progress-meters
+    // score of 50: `stroke-dasharray: 176 352`;
+    /** @type {?SVGCircleElement} */
+    const gaugeArc = gauge.querySelector('.lh-gauge-arc');
+    if (gaugeArc) {
+      gaugeArc.style.strokeDasharray = `${numericScore * 352} 352`;
+    }
+
+    const scoreOutOf100 = Math.round(numericScore * 100);
+    const percentageEl = this._shadowRoot.querySelector('.lh-gauge__percentage');
+    if (!percentageEl) throw new Error('Impossible');
+    percentageEl.textContent = scoreOutOf100.toString();
+    if (this._attributes.score === 'null') {
+      percentageEl.textContent = '?';
+      // @ts-ignore
+      percentageEl.title = 'Error';
+    }
+
+    const labelEl = this._shadowRoot.querySelector('.lh-gauge__label');
+    if (!labelEl) throw new Error('Impossible');
+    labelEl.textContent = this._attributes.label;
+  }
+}
+
+// Allow Node require()'ing.
+if (typeof module !== 'undefined' && module.exports) {
```

Comment:
When is this even going to be loaded in Node? Extending HTMLElement will break without a DOM emulation library I'm aware of (jsdom still doesn't support custom elements for reasons). Can you not use modules?

Reply:
it won't be run in node, only for bundling

Reply:
(also honestly just a copy of the pattern that exists in every single one of our files, I don't even know if this particular structure works for bundling)

Reply:
If you're using a bundler (Rollup, I hope!) `import`s will likely be much easier for it to understand.